### DISCUSSION TITLE
Format flux components with prettier if configured

### DISCUSSION THREAD

**Discussion by viceice** [renovatebot/renovate#23314] [Suggest an Idea]
### Tell us more.

Currently i've changes like this

![image](https://github.com/renovatebot/renovate/assets/1798109/830af5d3-0ccd-4221-8e04-bef3850d0f74)

So i would prefer to run prettier on those files, so ci checks won't fail.
We already have prettier as dependency, so should be easy to integrate in flux manager.

@rarkins  @secustor  WDYT?

**Comment by rarkins** [comment]
So you have text in your repo which needs reformatting?

**Comment by secustor** [comment]
I do not see why we shouldn't add `prettier` as a postUpdateOption.

**Reply by rarkins** [reply]
Previously when I looked at it, there was a challenge detecting which config/params to use as it wasn't standardized. Eg common to use CLI params buried in a package.json

**Reply by viceice** [reply]
We have a function to apply prettier format, i just want to call it as postUpdateOption. I'll try it on a reproduction and see what happens.

https://github.com/renovatebot/renovate/blob/5eeb2dfb3386522208ce9795307a4dd8a41e42b5/lib/workers/repository/config-migration/branch/migrated-data.ts#L39

**Reply by rarkins** [reply]
Do you mean prettier formatting just to files we update?

If it's prettier formatting to other files on the file system then the challenge is how to know which file patterns the repo wants prettier'd

**Reply by secustor** [reply]
I would assume there would be a prettier config in the repo to define this, at least in the first iteration.

**Reply by rarkins** [reply]
I don't think you can make that assumption. For example with this repo, we have no single source of truth "prettier config", you have to look at https://github.com/renovatebot/renovate/blob/2092254ce3c23d33201491cbbc0633dae44ba1b0/package.json#L42 to know which file patterns we require. Unless prettier has since standardized putting file patterns in the `.prettierrc`, there is no standard or convention for discovering this (we'd have to use some fuzzy logic to achieve reasonable accuracy or define our own convention)

**Reply by viceice** [reply]
I would only format the files which we've changed. I like the idea of a new post upgrade command.

**Reply by HonkingGoose** [reply]
## Prettier conventions

> Unless prettier has since standardized putting file patterns in the `.prettierrc`, there is no standard or convention for discovering this

Here are the relevant quotes from the Prettier docs, plus links:

- [Prettier config file locations](https://prettier.io/docs/en/configuration.html)
- [Prettier file patterns](https://prettier.io/docs/en/cli.html#file-patterns)

### Config file locations

> Prettier uses [cosmiconfig](https://github.com/davidtheclark/cosmiconfig) for configuration file support. This means you can configure Prettier via (in order of precedence):
> 
> - A `"prettier"` key in your `package.json` file.
> - A `.prettierrc` file written in JSON or YAML.
> - A `.prettierrc.json`, `.prettierrc.yml`, `.prettierrc.yaml`, or `.prettierrc.json5` file.
> - A `.prettierrc.js`, or `prettier.config.js` file that exports an object using `export default` or `module.exports` (depends on the [`type`](https://nodejs.org/api/packages.html#type) value in your `package.json`).
> - A `.prettierrc.mjs`, or `prettier.config.mjs` file that exports an object using `export default`.
> - A `.prettierrc.cjs`, or `prettier.config.cjs` file that exports an object using `module.exports`.
> - A `.prettierrc.toml` file.
> 
> The configuration file will be resolved starting from the location of the file being formatted, and searching up the file tree until a config file is (or isn’t) found.
> 
> Prettier intentionally doesn’t support any kind of global configuration. This is to make sure that when a project is copied to another computer, Prettier’s behavior stays the same. Otherwise, Prettier wouldn’t be able to guarantee that everybody in a team gets the same consistent results.
> 
> The options you can use in the configuration file are the same as the [API options](https://prettier.io/docs/en/options.html).

### File patterns

> Given a list of paths/patterns, the Prettier CLI first treats every entry in it as a literal path.
> 
> - If the path points to an existing file, Prettier CLI proceeds with that file and doesn’t resolve the path as a glob pattern.
> 
> - If the path points to an existing directory, Prettier CLI recursively finds supported files in that directory. This resolution process is based on file extensions and well-known file names that Prettier and its [plugins](https://prettier.io/docs/en/plugins.html) associate with supported languages.
> 
> - Otherwise, the entry is resolved as a glob pattern using the [glob syntax from the `fast-glob` module](https://github.com/mrmlnc/fast-glob#pattern-syntax).
> 
> Prettier CLI will ignore files located in `node_modules` directory. To opt out from this behavior, use `--with-node-modules` flag.
> 
> Prettier CLI will not follow symbolic links when expanding arguments.
> 
> To escape special characters in globs, one of the two escaping syntaxes can be used: `prettier "\[my-dir]/*.js"` or `prettier "[[]my-dir]/*.js"`. Both match all JS files in a directory named `[my-dir]`, however the latter syntax is preferable as the former doesn’t work on Windows, where backslashes are treated as path separators.

## Fuzzy logic or define own convention

> (we'd have to use some fuzzy logic to achieve reasonable accuracy or define our own convention)

Fuzzy logic is going to be tricky, with that many places to put the Prettier config. Defining our own convention seems easier, like this:

> ### Using Prettier with Renovate
>
> To use [Prettier](https://prettier.io/) with Renovate you must follow these rules.
>
> Only use these Prettier configuration file names and locations:
>
> - Prettier config file 1 that you want to support
> - Any other Prettier config files you want to support
>
> Renovate ignores all other Prettier configuration files or locations.
>
> Any Prettier file patterns must follow these rules:
>
> - Prettier file pattern number 1
> - Any other Prettier file patterns you want to support
>
> Renovate ignores all other Prettier file patterns.

**Reply by viceice** [reply]
interesting, but I don't want to format all files. I only like to format the files generated by flux cli call. especially only the flux system components file.

**Reply by secustor** [reply]
That could be a separate option. 
If enabled we can capture the `git diff`, apply prettier and only commit the files from the diff.
